### PR TITLE
ShortcutExpander has been extended in a way that it will examine a ha…

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What this PR does / why we need it**:

ShortcutExpander has been extended in a way that it will examine a hardcoded list of tuples anticipated from the server when searching for an alternative name for the resource.

Note that the list is ordered and the first match will yield the extended resource's name.

One important thing to highlight is that the ShortcutExpander will fall back to PriorityRestMaper to determine the group for the resource.

Also this PR introduces a new shortcut namely sc which will resolve to storageclasses within storage.k8s.io group

**Special notes for your reviewer**: You might want to see https://github.com/kubernetes/kubernetes/pull/38755

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by fabianofranz** [comment]
@k8s-bot ok to test

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 1]
```diff
@@ -124,14 +124,55 @@ func (e ShortcutExpander) AliasesForResource(resource string) ([]string, bool) {
...
 	}
-	expanded := expandResourceShortcut(schema.GroupVersionResource{Resource: resource}).Resource
+	expanded := e.expandResourceShortcut(schema.GroupVersionResource{Resource: resource}).Resource
 	return []string{expanded}, (expanded != resource)
 }
 
+// resourceShortcuts represents a structure that holds the information how to
+// transition from resource's shortcut to its full name.
+type resourceShortcuts struct {
+	from schema.GroupResource
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: shortForm

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 2]
```diff
@@ -124,14 +124,55 @@ func (e ShortcutExpander) AliasesForResource(resource string) ([]string, bool) {
...
+	to   schema.GroupResource
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: longForm

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 3]
```diff
@@ -124,14 +124,55 @@ func (e ShortcutExpander) AliasesForResource(resource string) ([]string, bool) {
...
+}
+
+// getServerRes returns a hardcoded set of tuples. Note that the list
+// is ordered by group priority.
+func (e ShortcutExpander) getServerRes() ([]resourceShortcuts, error) {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: Add a TODO to wire this up to discovery.

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 4]
```diff
@@ -124,14 +124,55 @@ func (e ShortcutExpander) AliasesForResource(resource string) ([]string, bool) {
...
+				if resource.Resource == item.from.Resource {
+					resource.Resource = item.to.Resource
+					return resource
+				}
+			}
+		}
+	}
+
+	// fall back to the hardcoded list
 	if expanded, ok := kubectl.ShortForms[resource.Resource]; ok {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: Let's rework these defaults to the be the same format as the `getServerRes` method.  Then have the evaluation from inside the if extracted to a method.

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 3 above)
Comment (deads2k) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: `getShortcutMappings`

**Comment by deads2k** [comment]
Minor comments.  It's looking good.

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 4 above)
Comment (p0lyn0mial) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: Alright, can I do this in a separate pull request ?

Comment (deads2k) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: > Alright, can I do this in a separate pull request ?

I'm afraid I see them as tightly coupled.  You could just have the method return the current shortcuts in the corrected format.  If we only collect the shortcuts once (possible), then it factors very cleanly to have these at the very bottom of the list.

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 5]
```diff
@@ -124,17 +124,57 @@ func (e ShortcutExpander) AliasesForResource(resource string) ([]string, bool) {
...
+func (e ShortcutExpander) getShortcutMappings() ([]resourceShortcuts, error) {
+	res := []resourceShortcuts{
+		{
+			shortForm: schema.GroupResource{Group: "storage.k8s.io", Resource: "sc"},
+			longForm:  schema.GroupResource{Group: "storage.k8s.io", Resource: "storageclasses"},
+		},
+	}
+
+	// append hardcoded short forms at the end of the list
+	for short, long := range kubectl.ShortForms {
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: I think that `kubectl.ShortForms` should be a `resourceShortcut` or at least the same shape.  If you aren't familiar enough to know all the groups for the transform, I can do it in a followup.

Comment (p0lyn0mial) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: ah, sorry for this misunderstanding. Actually that what my initial idea - that's why I wanted to have a separate pull for that. But after your previous comments I thought that you want only wrap it in a function.

Comment (p0lyn0mial) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: alright I will change the type of `kubectl.ShortForms`, adjust the existing code and then it would be nice if you could fill the groups info

File: pkg/kubectl/kubectl.go [hunk 6]
```diff
@@ -75,42 +75,117 @@ func (m OutputVersionMapper) RESTMapping(gk schema.GroupKind, versions ...string
...
+	{
+		ShortForm: schema.GroupResource{Resource: "cm"},
+		LongForm:  schema.GroupResource{Resource: "configmaps"},
+	},
+	{
+		ShortForm: schema.GroupResource{Resource: "cs"},
+		LongForm:  schema.GroupResource{Resource: "componentstatuses"},
+	},
+	{
+		ShortForm: schema.GroupResource{Resource: "csr"},
```
Comment (deads2k) [inline pkg/kubectl/kubectl.go]: certificates.k8s.io

File: pkg/kubectl/kubectl.go [hunk 7]
```diff
@@ -75,42 +75,117 @@ func (m OutputVersionMapper) RESTMapping(gk schema.GroupKind, versions ...string
...
+		LongForm:  schema.GroupResource{Resource: "certificatesigningrequests"},
+	},
+	{
+		ShortForm: schema.GroupResource{Resource: "deploy"},
+		LongForm:  schema.GroupResource{Resource: "deployments"},
+	},
+	{
+		ShortForm: schema.GroupResource{Resource: "ds"},
```
Comment (deads2k) [inline pkg/kubectl/kubectl.go]: extensions

File: pkg/kubectl/kubectl.go [hunk 8]
```diff
@@ -75,42 +75,117 @@ func (m OutputVersionMapper) RESTMapping(gk schema.GroupKind, versions ...string
...
+	{
+		ShortForm: schema.GroupResource{Resource: "ev"},
+		LongForm:  schema.GroupResource{Resource: "events"},
+	},
+	{
+		ShortForm: schema.GroupResource{Resource: "hpa"},
+		LongForm:  schema.GroupResource{Resource: "horizontalpodautoscalers"},
+	},
+	{
+		ShortForm: schema.GroupResource{Resource: "ing"},
```
Comment (deads2k) [inline pkg/kubectl/kubectl.go]: extensions

File: pkg/kubectl/kubectl.go [hunk 9]
```diff
@@ -75,42 +75,117 @@ func (m OutputVersionMapper) RESTMapping(gk schema.GroupKind, versions ...string
...
+	{
+		ShortForm: schema.GroupResource{Resource: "no"},
+		LongForm:  schema.GroupResource{Resource: "nodes"},
+	},
+	{
+		ShortForm: schema.GroupResource{Resource: "ns"},
+		LongForm:  schema.GroupResource{Resource: "namespaces"},
+	},
+	{
+		ShortForm: schema.GroupResource{Resource: "pdb"},
```
Comment (deads2k) [inline pkg/kubectl/kubectl.go]: policy

File: pkg/kubectl/kubectl.go [hunk 10]
```diff
@@ -75,42 +75,117 @@ func (m OutputVersionMapper) RESTMapping(gk schema.GroupKind, versions ...string
...
+		LongForm:  schema.GroupResource{Resource: "poddisruptionbudgets"},
+	},
+	{
+		ShortForm: schema.GroupResource{Resource: "po"},
+		LongForm:  schema.GroupResource{Resource: "pods"},
+	},
+	{
+		ShortForm: schema.GroupResource{Resource: "psp"},
```
Comment (deads2k) [inline pkg/kubectl/kubectl.go]: extensions

File: pkg/kubectl/kubectl.go [hunk 11]
```diff
@@ -75,42 +75,117 @@ func (m OutputVersionMapper) RESTMapping(gk schema.GroupKind, versions ...string
...
-	"psp":    "podSecurityPolicies",
-	"pvc":    "persistentvolumeclaims",
-	"pv":     "persistentvolumes",
-	"quota":  "resourcequotas",
-	"rc":     "replicationcontrollers",
-	"rs":     "replicasets",
-	"sa":     "serviceaccounts",
-	"svc":    "services",
+
+	// TODO (deads2k): Please add group information and prioritise by group if applicable.
```
Comment (deads2k) [inline pkg/kubectl/kubectl.go]: 1. no group comes first (those are legacy ones)
1. autoscaling
1. certificates
1. policy
1. extensions

**Comment by deads2k** [comment]
make the group updates, order, and squash and this lgtm.  We can get the rest in followups.

**Comment by deads2k** [comment]
@k8s-bot ok to test

**Comment by deads2k** [comment]
@p0lyn0mial thanks for picking this up, it sets up for discoverable resource short names quite well.

**Comment by deads2k** [comment]
lgtm, but we an unforeseen snag.

https://github.com/kubernetes/kubernetes/blob/master/hack/make-rules/test-cmd.sh#L1076-L1077  It's a prefix matching test where partial groups are assumed.  I think its just a prefix match in this code.

**Comment by p0lyn0mial** [comment]
I think it boils down to implementing something similar to https://github.com/kubernetes/kubernetes/blob/master/pkg/api/meta/restmapper.go#L258

Do you want to add support for prefixes in this PR ?

Why test-cmd.sh is not part of testing ?

**Comment by deads2k** [comment]
> Do you want to add support for prefixes in this PR ?

Yes, otherwise we'll regress `kubectl` in master.



> Why test-cmd.sh is not part of testing ?

Not sure.  `hack/test-cmd.sh` gets the job done locally.  I wouldn't worry about it, we have CI for a reason.

**Comment by deads2k** [comment]
lgtm

**Comment by deads2k** [comment]
@k8s-bot test this
